src/PowerShellEditorServices/Services/DebugAdapter/Handlers/StackTraceHandler.cs

```diff
@@ -1,64 +1,132 @@
+        }
 
-        public StackTraceHandler(DebugService debugService) => _debugService = debugService;
+        // Adapting to int to let us use LINQ, realistically if you have a stacktrace larger than this that the client is requesting, you have bigger problems...
+        int skip = Convert.ToInt32(request.StartFrame ?? 0);
+        int take = Convert.ToInt32(request.Levels ?? 0);
 
-        public async Task<StackTraceResponse> Handle(StackTraceArguments request, CancellationToken cancellationToken)
-        {
-            StackFrameDetails[] stackFrameDetails = await _debugService.GetStackFramesAsync(cancellationToken).ConfigureAwait(false);
+        // We generate a label for the breakpoint and can return that immediately if the client is supporting DelayedStackTraceLoading.
+        InvocationInfo invocationInfo = debugService.CurrentDebuggerStoppedEventArgs?.OriginalEvent?.InvocationInfo
+            ?? throw new NotSupportedException("InvocationInfo was not available on CurrentDebuggerStoppedEvent args. This is a bug.");
 
-            // Handle a rare race condition where the adapter requests stack frames before they've
-            // begun building.
-            if (stackFrameDetails is null)
-            {
-                return new StackTraceResponse
-                {
-                    StackFrames = Array.Empty<StackFrame>(),
-                    TotalFrames = 0
-                };
-            }
-
-            List<StackFrame> newStackFrames = new();
-
-            long startFrameIndex = request.StartFrame ?? 0;
-            long maxFrameCount = stackFrameDetails.Length;
-
-            // If the number of requested levels == 0 (or null), that means get all stack frames
-            // after the specified startFrame index. Otherwise get all the stack frames.
-            long requestedFrameCount = request.Levels ?? 0;
-            if (requestedFrameCount > 0)
-            {
-                maxFrameCount = Math.Min(maxFrameCount, startFrameIndex + requestedFrameCount);
-            }
+        StackFrame breakpointLabel = CreateBreakpointLabel(invocationInfo);
 
-            for (long i = startFrameIndex; i < maxFrameCount; i++)
+        if (skip == 0 && take == 1) // This indicates the client is doing an initial fetch, so we want to return quickly to unblock the UI and wait on the remaining stack frames for the subsequent requests.
+        {
+            return new StackTraceResponse()
             {
-                // Create the new StackFrame object with an ID that can
-                // be referenced back to the current list of stack frames
-                newStackFrames.Add(LspDebugUtils.CreateStackFrame(stackFrameDetails[i], id: i));
-            }
+                StackFrames = new StackFrame[] { breakpointLabel },
+                TotalFrames = INITIAL_PAGE_SIZE //Indicate to the client that there are more frames available
```

Comment:
```suggestion
                TotalFrames = INITIAL_PAGE_SIZE // Indicate to the client that there are more frames available
```

Does this have any effect on the client's UI for different values of `TotalFrames`? I ask since we arbitrarily set it to 20.

Reply:
No, according to the spec[ `a client should be prepared to receive fewer frames than requested, which is an indication that the end of the stack has been reached.`](https://microsoft.github.io/debug-adapter-protocol/specification#Requests_StackTrace)